src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -128,16 +133,150 @@ static QueryModelVisitor()
+						// Polymorphic case complex to handle and not implemented. (HQL query must be reshaped for adding
+						// additional data to allow a meaningful overall average computation.)
+						// Leaving it untouched for allowing non polymorphic cases to work.
+						break;
+					case NhExpressionType.Count:
+						AddPostExecuteTransformerForCount();
+						break;
+					case NhExpressionType.Max:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.EnumerableMethods.MaxDefinition);
+						break;
+					case NhExpressionType.Min:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.EnumerableMethods.MinDefinition);
+						break;
+					case NhExpressionType.Sum:
+						AddPostExecuteTransformerForSum();
+						break;
+				}
+			}
+		}
+
+		private void AddPostExecuteTransformerForCount()
+		{
+			// Count results have to be summed. No null case to take into account.
+			var elementType = Model.ResultTypeOverride;
+			var inputListType = typeof(IQueryable<>).MakeGenericType(elementType);
+			var inputList = Expression.Parameter(inputListType, "inputList");
+			// Sum has no suitable generic overload, throw in Sum on int, then the code using it
+			// will check and adjust it if it is long instead of int (GetAggregateMethodCall does that).
+			var aggregateCall = GetAggregateMethodCall(ReflectionCache.EnumerableMethods.SumOnInt, inputListType, elementType, inputList);
+			_hqlTree.AddPostExecuteTransformer(Expression.Lambda(aggregateCall, inputList));
+		}
+
+		private void AddPostExecuteTransformerForResultAggregate(MethodInfo aggregateMethodTemplate)
+		{
+			var elementType = Model.ResultTypeOverride;
+
+			LambdaExpression aggregateLambda;
+
+			// string may be aggregated and are not Nullable<> but can be null, and qualifies as reference
+			if (elementType.IsNullableOrReference())
+			{
+				var inputListType = typeof(IQueryable<>).MakeGenericType(elementType);
+				var inputList = Expression.Parameter(inputListType, "inputList");
+				var aggregateCall = GetAggregateMethodCall(aggregateMethodTemplate, inputListType, elementType, inputList);
+				aggregateLambda = Expression.Lambda(aggregateCall, inputList);
+			}
+			else
+			{
+				// On non nullable, we have to preserve the behavior "nothing to aggregate => exception",
+				// while supporting the polymorphic case "nothing to aggregate for some classes but something to aggregate
+				// for others" => aggregate of others.
+				// For this, overriding result type of concrete queries as nullable, aggregating concrete results,
+				// casting back to non nullable (and failing if null).
+				// This still causes a change: the expected failure case will yield an InvalidOperationException instead of
+				// a GenericADOException with an ArgumentNullException inner exception.
+
+				var nullableElementType = typeof(Nullable<>).MakeGenericType(new[] { elementType });
+				var nullableInputListType = typeof(IQueryable<>).MakeGenericType(nullableElementType);
```

Comment:
Reviewing again some other parts of the fix, it appears to me I was not consistent with that. This fix has spanned a long period, I have ended up not remembering all of it.